python/ray/experimental/workflow/storage/s3.py

```diff
@@ -0,0 +1,218 @@
+import functools
+import tempfile
+import json
+import aioboto3
+import itertools
+import ray
+from typing import Any, Awaitable, Dict, Callable
+from ray.experimental.workflow.common import StepID
+from ray.experimental.workflow.storage.base import (
+    Storage, ArgsType, StepStatus, DataLoadError, DataSaveError)
+# constants used in filesystem
+OBJECTS_DIR = "objects"
+STEPS_DIR = "steps"
+STEP_INPUTS_METADATA = "inputs.json"
+STEP_OUTPUTS_METADATA = "outputs.json"
+STEP_ARGS = "args.pkl"
+STEP_OUTPUT = "output.pkl"
+STEP_FUNC_BODY = "func_body.pkl"
+
+MAX_RECEIVED_DATA_MEMORY_SIZE = 25 * 1024 * 1024  # 25MB
+
+
+def data_load_error(func):
+    @functools.wraps(func)
+    async def _func(*args, **kvargs):
+        try:
+            ret = await func(*args, **kvargs)
+            return ret
+        except Exception as e:
+            raise DataLoadError from e
+
+    return _func
+
+
+def data_save_error(func):
+    @functools.wraps(func)
+    async def _func(*args, **kv_args):
+        try:
+            ret = await func(*args, **kv_args)
+            return ret
+        except Exception as e:
+            raise DataSaveError from e
+
+    return _func
+
+
+class S3StorageImpl(Storage):
```

Comment:
Any reason to use boto directly instead of smart_open?

Reply:
smart_open doesn't support async, operations. aioboto can read data in parallel with async API. It's mainly for performance concerns. From the previous experiment, it does matter a lot here.